### ISSUE TITLE
sql: we need to audit recent new uses of InternalExecutor, especially iterator, for invalid uses of kv.Txn

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#99211]
TLDR: we need to audit the code to verify uses of InternalExecutor, Iterator and Txn -- especially those introduced through the last 6 months.

See issue #99209 for context.

Given the context in #99209 and #99171 (and this [internal convo](https://cockroachlabs.slack.com/archives/C0168LW5THS/p1679437525752149)) we suspect that we have introduced bugs inadvertently.

- They may explain an uptick in test flakes in the past few weeks.
- There is no reliable "assertion" in the code that detects the problem when it happens.
- They will, if encountered in prod, create hard-to-troubleshoot stability and correctness issues.

Through the v23.1 cycle, multiple teams have started to use InternalExecutor, its iterator pattern, or expanded its use.
Including but not limited to:
- in Jobs
- In Schema Changer
- in SQL / KV Observability (in API endpoints and virtual tables)

It's possible (and we estimate likely) that the folk in these teams were _not_ aware of the API limitation outlined in #99209 and thus have introduced these "invisible" bugs inadvertently.

To make progress this issue proposes a two-pronged approach:

1. audit the code, especially recent changes / new uses of internalexecutor and its iterator, to see if there are potential Txn API violations as described in #99209.
2. concurrently, work on an API change inside kv.Txn and InternalExecutor to add an assertion that detects invalid API use and reports errors when it happens, then run our CI with this.

The audit is necessary even if we do point (2) because we know certain combinations of calls are not currently exercised in tests.

Jira issue: CRDB-25760

**Comment by knz** [comment]
Summary from the job meeting today - we have a proposal that avoids a full code audit.

1. work towards fixing #99171 in any case
2. turn off distsql by default in internal executor (possibly restricted to QueryIterator - TBD) and then selectively re-enable it only in the handful of scale-sensitive IE queries (constraint validation) where safety can be more narrowly audited. 
3. additional assertion check inside RootTxn / LeafTxn to detect concurrent/simultaneous use and panic (at least in tests)
4. Change the error handling to avoid panics caused by assertion failures in prod clusters (to avoid rolling cluster failures). In particular turn the KV assertions from panics to errors.

**Comment by knz** [comment]
@arulajmani :
> -  I don’t think turning off distsql goes all the way — we’d also need to disable the Streamer here to ensure LeafTxns aren’t used.
> - changing these fatals to an assertion failed came up in the KV On-Call weekly as well. I’ve got the action item to file an issue and write up the patch.

**Comment by arulajmani** [comment]
Removing the release blocker status on this issue now that we've backported https://github.com/cockroachdb/cockroach/pull/99328 to 23.1 and we no longer crash nodes. This issue continues to remain a GA-blocker though.

**Comment by msirek** [comment]
Assigning @michae2 to audit the most recent additions of use of the internal execution in SQL Queries code.

We'd like some help from SQL Foundations for other portions.

**Comment by michae2** [comment]
@knz do we still need to do the audit now that https://github.com/cockroachdb/cockroach/pull/99325 has merged?

**Comment by knz** [comment]
let's keep this one sleep now.
